pkg/operator/controllers/machine_controller.go

```diff
@@ -0,0 +1,252 @@
+	}
+
+	if machine.Spec.ProviderSpec.Value == nil {
+		return true, nil, fmt.Errorf("provider spec is missing in the machine %q", machine.Name)
+	}
+
+	o, _, err := scheme.Codecs.UniversalDeserializer().Decode(machine.Spec.ProviderSpec.Value.Raw, nil, nil)
+	if err != nil {
+		return true, nil, err
+	}
+
+	machineProviderSpec, ok := o.(*azureproviderv1beta1.AzureMachineProviderSpec)
+	if !ok {
+		// This should never happen: codecs uses scheme that has only one registered type
+		// and if something is wrong with the provider spec - decoding should fail
+		return true, nil, fmt.Errorf("failed to read provider spec from the machine %q: %T", machine.Name, o)
+	}
+
+	if !vmSizeIsValid(api.VMSize(machineProviderSpec.VMSize), r.developmentMode, isMaster) {
+		supported = false
+		msgs = append(msgs, fmt.Sprintf("the machine %s VM size '%s' is invalid", machine.Name, machineProviderSpec.VMSize))
+	}
+
+	if machineProviderSpec.OSDisk.DiskSizeGB < 128 {
+		supported = false
+		msgs = append(msgs, fmt.Sprintf("the machine %s disk size '%d' is invalid", machine.Name, machineProviderSpec.OSDisk.DiskSizeGB))
+	}
+
+	// to begin with, just check that the image publisher and offer are correct
+	if machineProviderSpec.Image.Publisher != "azureopenshift" || machineProviderSpec.Image.Offer != "aro4" {
+		supported = false
+		msgs = append(msgs, fmt.Sprintf("the machine %s image '%v' is invalid", machine.Name, machineProviderSpec.Image))
+	}
+
+	if machineProviderSpec.ManagedIdentity != "" {
+		supported = false
+		msgs = append(msgs, fmt.Sprintf("the machine %s managedIdentity '%v' is invalid", machine.Name, machineProviderSpec.ManagedIdentity))
+	}
+
+	return supported, msgs, nil
+}
+
+func (r *MachineReconciler) checkMachines(ctx context.Context) (bool, []string, error) {
+	msgs := []string{}
+	supported := true
+	actualWorkers := 0
+	actualMasters := 0
+
+	machines, err := r.clustercli.MachineV1beta1().Machines(machineSetsNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return supported, msgs, err
+	}
+	for _, machine := range machines.Items {
+		mSupported, msgsMachine, err := r.machineSupported(ctx, &machine)
+		if err != nil {
+			r.log.Errorf("machinesSupported err:%v", err)
+			return supported, msgs, err
+		}
+		if !mSupported {
+			supported = false
+			msgs = append(msgs, msgsMachine...)
+		}
+		isMaster, err := machineIsMaster(&machine)
+		if err != nil {
+			return supported, msgs, err
+		}
+		if isMaster {
+			actualMasters++
+		} else {
+			actualWorkers++
+		}
+	}
+
+	expectedMasters := 3
+	expectedWorkers, err := r.workerReplicas()
+	if err != nil {
+		return supported, msgs, err
+	}
+
+	if actualMasters != expectedMasters {
+		supported = false
+		msgs = append(msgs, fmt.Sprintf("invalid number of master machines %d", actualMasters))
+	}
+
+	if actualWorkers != expectedWorkers {
+		supported = false
+		msgs = append(msgs, fmt.Sprintf("invalid number of worker machines %d, expected %d", actualWorkers, expectedWorkers))
+	}
```

Comment:
this is proving that the number of worker machines == the number specified in the machinesets, aka a smoke test of the machineset controller?  Is this needed?  Is it not already covered elsewhere?

Reply:
so we had a customer that created a machine not in the context of a machineset. they basically copy/pasted the output of a machine yaml and created it (with changes). i am trying to catch that..